torch/distributed/fsdp/_init_utils.py

```diff
@@ -22,6 +55,219 @@
 FSDP_SYNCED = "_fsdp_synced"
 
+# TODO (awgu): Refactor this later
+SHARDING_STRATEGY_MAP = {
+    ShardingStrategy.NO_SHARD: HandleShardingStrategy.NO_SHARD,
+    ShardingStrategy.FULL_SHARD: HandleShardingStrategy.FULL_SHARD,
+    ShardingStrategy.SHARD_GRAD_OP: HandleShardingStrategy.SHARD_GRAD_OP,
+}
+
+
+# NOTE: Since non-self attributes cannot be type annotated, several attributes
+# on `state` are defined first as local variables before being assigned.
+
+
+@no_type_check
+def _init_process_group_state(
+    state: _State,
+    process_group: Optional[dist.ProcessGroup],
+) -> _State:
+    state.process_group = process_group or _get_default_group()
+    state.rank = state.process_group.rank()
+    state.world_size = state.process_group.size()
+    return state
+
+
+@no_type_check
+def _init_ignored_module_states(
+    state: _State,
+    module: nn.Module,
+    ignored_modules: Optional[Iterable[torch.nn.Module]],
+) -> _State:
+    state._ignored_modules = _get_ignored_modules(module, ignored_modules)
+    state._ignored_params, state._ignored_param_names = _get_ignored_params(
+        module,
+        state._ignored_modules,
+    )
+    # TODO: FSDP's contract for buffers is not well-defined. They are
+    # implicitly ignored for most functionality since they are not sharded;
+    # however, FSDP still imposes some semantics on buffers (e.g. buffer mixed
+    # precision). We should formalize this contract and decide if we need to
+    # compute and store `_ignored_buffers`.
+    return state
+
+
+@no_type_check
+def _init_module_and_device_state(
+    state: _State,
+    module: nn.Module,
+    device_id: Optional[Union[int, torch.device]],
+    param_init_fn: Optional[Callable[[nn.Module], None]],
+    sync_module_states: bool,
+) -> _State:
+    state._buffer_names = _get_buffer_names(module)
+    # Save a mapping from fully prefixed buffer name to its original dtype
+    # since when buffer mixed precision is enabled, buffers are restored to
+    # their original dtype for model checkpointing
+    _buffer_name_to_orig_dtype: Dict[str, torch.dtype] = {}
+    state._buffer_name_to_orig_dtype = _buffer_name_to_orig_dtype
+    _check_single_device_module(module, state._ignored_params)
+    device_from_device_id = _get_device_from_device_id(device_id, state.rank)
+    _materialize_module(
+        module,
+        param_init_fn,
+        state._ignored_params,
+        device_from_device_id,
+        lambda _: True,
+    )
+    # TODO: We need to skip this for functional-like to avoid moving the entire
+    # unsharded module onto GPU before any sharding.
+    _move_module_to_device(module, state._ignored_params, device_from_device_id)
+    state.compute_device = _get_compute_device(
+        module,
+        state._ignored_params,
+        device_from_device_id,
+        state.rank,
+    )
+    state._managed_params = list(_get_orig_params(module, state._ignored_params))
+    if sync_module_states:
```

Comment:
when we don't need to sync states? loaded from previous checkpoint? Any other cases?

Reply:
Also, I think we do not need to sync if we assume that the model initialization is deterministically the same across ranks.